### ISSUE TITLE
[34] Sparse is Enough in Scaling Transformers (Terraformer)

### ISSUE THREAD

**Issue by dhkim0225** [dhkim0225/1day_1paper#62]
[paper](https://arxiv.org/pdf/2111.12763.pdf)
[code](https://github.com/google/trax)

inference 할 때 sparsity 적용해서 가속.
1. relu activation 에서 0인 부분을 예측하는 모듈 추가. (sparse feed-forward)
2. qkv linear 를 permutation + conv 로 대체. 

17B 모델에서 토큰 당 3.7초 걸리던 것을 0.1 ~ 0.2초 수준으로 낮췄음. (cpu, batch size 1)

![image](https://user-images.githubusercontent.com/16400591/144162116-bedfa636-5954-4678-852f-38d84e5ae2ea.png)

# INTRO
decoder 구조는 3가지 부분으로 나눌 수 있다.
1. masked-self-attention
2. cross-attention
3. feed-forward

computation을 따져보자.
BERT 의 <img src="https://render.githubusercontent.com/render/math?math=d_{model} = 768"> 
GPT-3 의 <img src="https://render.githubusercontent.com/render/math?math=d_{model} = 12288"> 
이다. 

masked-self-attention 에서 q, k, v 에 weight 곱해주고, output 내보낼 때도 weight 곱해주니까 computation 은 <img src="https://render.githubusercontent.com/render/math?math=4 d_{model}^2"> 이다.
cross-attention 에서는, encoder output 으로 key-value 는 cache 되어 있으니까 위 예시처럼 4개가 아닌, 2개만 활성화된다. <img src="https://render.githubusercontent.com/render/math?math=2 d_{model}^2">
feed-forward 는 <img src="https://render.githubusercontent.com/render/math?math=d_{model}d_{ff}"> 보통 feed-forward 크기를 model-dimension 에 4를 곱해서 사용하니까, 이것도 상당히 크다.
총 computation 은 <img src="https://render.githubusercontent.com/render/math?math=4 d_{model}^2 %2B\ 2 d_{model}^2 %2B\ d_{model}d_{ff}"> 이다.
이걸 압착시키는 테크닉을 제안한다. 최종 computation은 이렇게 변한다. <img src="https://render.githubusercontent.com/render/math?math=8 d_{model}^{1.5} %2B\ 4 d_{model}^{1.5} %2B\ 4 d_{model}^{1.5}">

# Sparse is Enough
## Sparse Feedforward Layer
![image](https://user-images.githubusercontent.com/16400591/144170712-543a2db4-853b-4b45-8272-98ccb3a8120d.png)
그림이 너무 깔끔하다.

(b) 부터 살펴보자. 요 녀석은 controller 이다.
controller 는 dynamic 하게 activate 시킬 column을 골라준다. block 마다 (block 크기는 hyperparameter 이다) 단 하나의 column 만 activation 시켜 주도록 하는 것이다. 
feed-forward <img src="https://render.githubusercontent.com/render/math?math=d_{model}d_{ff}"> computation 이 크니까, <img src="https://render.githubusercontent.com/render/math?math=d_{ff}"> 의 크기를 어떻게든 줄이는 느낌.
그림에 적혀있는 `low-rank` 또한 hyperparameter 이다.
fc 2개를 거쳐서 0, 1 vector 를 내뱉어 주는데, argmax 쓰면 gradient 끊기니까 gumbell softmax 활용.

(a) 를 다음으로 살펴본다.
controller output 을 이용해서 어떻게 computation 하는 지 보여준다.
우선, d_ff 개의 vector를 뽑아내고 일종의 attention value 로 활용한다. 다만, controller output 을 곱해버려서 controller에서 허가(?) 한 녀석들만 attention에 활용한다.

실험적으로, T5-large 모델에서 `block 개수는 64`, `d_{low_rank} == 64` 로 설정했다.

## Sparse QKV Layer
![image](https://user-images.githubusercontent.com/16400591/144171556-87ea48fd-65f3-43af-95d4-775e0786712c.png)

이 논문. 그림 정말 맘에 든다.
(a) 는 Multiplicative layer 의 구조를 나타낸다.
(b) 는 sparse-QKV 를 conv 활용으로 구현했음을 보여준다.

아이디어는 단순하다. (a) multiplicative layer 는 SxM size 의 compact 한 attention output 을 뱉어주는 layer 이다. 이를 이용해서 (b) 에서 sparse 하게도 구성해 보고, mult 와 conv 를 함께 활용해서 sparse-QKV 를 구성도 해본 것이다.
multiplicative layer 에서 `M == d_{model} / S` 이다.

실험적으로, `conv filter F == 3`, `S == 16` 으로 설정했다고 한다.

아래 그림은 T-5 large 실험결과인데, variant 들끼리의 속도는 거의 비슷했다고 한다.
fair comparison 을 위해서 parameter 를 비슷하게 맞추고 진행했다고 한다.
feed-forward dimension 을 4096 이 아니라 6144 를 써서 param 크기를 맞췄다. ~(hmm.......)~
![image](https://user-images.githubusercontent.com/16400591/144174002-deecf014-a3f6-43b8-a81d-dc33e775f3f2.png)

승자는 mult variant !

## Sparse loss layer
final dense layer 또한, Multiplicative layer 를 활용해서 구성할 수 있다. 
논문에서는 dense 보다 computational-efficient 하게 동작하는 layer 정도로 Multiplicative layer 를 밀고 있기 때문이다.
last layer 에서의 S는 2를 쓰면 성능이 좋았... 지만, 속도가 거의 그대로.
S 가 늘어날 수록 multiplicative layer 에서의 M 은 줄어들어서 최종 output size 는 동일하다. (voca-size)
![image](https://user-images.githubusercontent.com/16400591/144173612-2f054a8e-2a2a-4e95-b1b2-84aba8c8c752.png)
![image](https://user-images.githubusercontent.com/16400591/144173629-9fef3a52-cd14-42c5-81fb-ff806e6ded37.png)

# Sparsity for Long Sequences (Terraformer)
sequence 가 길면 길수록 잘 동작을 안했고, 이를 위해 Terraformer 를 설계한다.
![image](https://user-images.githubusercontent.com/16400591/144176336-448a726e-90aa-4a6f-ba49-6d2da7dcf523.png)

## Architecture for Long Sequences
저자들은 Scaling Transformer 에 sparsity 를 넣다보니, decoder 단에서 self-attn, cross-attn 을 굳이 나눌 필요가 있나 생각이 들었다고 한다.
그래서 enc-dec cross attn 을 없애버리고, enc attn 을 함께 decoder 의 input으로 넣어주었다. 그리고 self-attn 을 하나 더 넣어주었더니, 속도도 빠르더라.

## Reversibility for Memory Efficiency
large batch training 을 위해 Reformer 에서도 적용했던, reversible 한 구조를 사용한다.
![image](https://user-images.githubusercontent.com/16400591/144179528-ad293167-3a19-4e36-8236-da45393f6194.png)

또, forward 할 때, discrete 한 decision들은 굉장히 중요하다고 한다. 어떤 row 부분을 선택할 것인지와 같은 discrete value 들을 cache 해놓고 backward 할 때 사용하는 것은 필수다.

## Recurrence for Generalization
generalization 을 위해 recurrent 모델을 사용했다. ~(insight는 잘 모르겠음)~
그래서 parallel execution 이 가능한 [SRU](https://arxiv.org/pdf/1709.02755.pdf) 를 사용했다.

SRU 내부는 dense layer 가 사용되기 때문에, sparse tensor 의 이점이 사라질 수 있다. 간단하게 sparse output 을 small dimension 으로 project 한 후 SRU를 통과시켰더니 괜찮았다고 한다. SRU 32 dimension을 사용했다고 한다.
C4 dataset 에서는 성능이 미미하지만, generalization 이 더 잘됨을 보여주기 위해 OOD 검출용 synthetic data 도 구성했다.

# Results
![image](https://user-images.githubusercontent.com/16400591/144181043-4bc199d6-a30b-476a-9133-70ddd0b43da3.png)
